Adds explicit "$schema" property to JSON Schema generated from Swagger

artem-zakharchenko opened the pull request:
## Changes

- Adds explicit `$schema` property in JSON Schema generated by OA2 document
- Adjusts `json-schema-test.js` to assert the presence of the root-level `$schema` property

## GitHub

- Fixes #399

artem-zakharchenko commented:
I'm currently experiencing failing tests under "Inherit Parameters" test. More details below.

During "Body Parameter" tests there is an assertion between the parsing result () and the actual data (taken from the source modified by `source.push(makeParameter())` before the relevant test):

https://github.com/apiaryio/api-elements.js/blob/3bbc63800fa41de7626453ff49bfd6aef2fdcf57/packages/fury-adapter-swagger/test/inherit-parameters-test.js#L235

The parsing result is valid, as it contains the added `$schema` property. The expected result from the source, however, doesn't contain that property, which makes assertion to fail. I'm looking at how the parameters are added via `makeParameter`:

https://github.com/apiaryio/api-elements.js/blob/3bbc63800fa41de7626453ff49bfd6aef2fdcf57/packages/fury-adapter-swagger/test/inherit-parameters-test.js#L46

This line is invoked in the first "Body Parameter" test when `makeParameter('test', 'body')` is pushed to the `source`. When I attempt to add the `$schema` property for two payloads to match I get a list of `Annotation` from Fury, indicating parsing errors:

```
[ Annotation {
    _content: 'Data does not match any schemas from \'oneOf\'',
    _storedElement: 'annotation',
    _meta:
     ObjectElement { _content: [Array], _storedElement: 'object' },
    _attributes:
     ObjectElement { _content: [Array], _storedElement: 'object' } },
  Annotation {
    _content: 'Source maps are only available with string input',
    _storedElement: 'annotation',
    _meta:
     ObjectElement { _content: [Array], _storedElement: 'object' },
    _attributes:
     ObjectElement { _content: [Array], _storedElement: 'object' } },
  Annotation {
    _content: 'Data does not match any schemas from \'oneOf\'',
    _storedElement: 'annotation',
    _meta:
     ObjectElement { _content: [Array], _storedElement: 'object' },
    _attributes:
     ObjectElement { _content: [Array], _storedElement: 'object' } },
  Annotation {
    _content: 'Missing required property: $ref',
    _storedElement: 'annotation',
    _meta:
     ObjectElement { _content: [Array], _storedElement: 'object' },
    _attributes:
     ObjectElement { _content: [Array], _storedElement: 'object' } },
  Annotation {
    _content: 'Additional properties not allowed: $schema',
    _storedElement: 'annotation',
    _meta:
     ObjectElement { _content: [Array], _storedElement: 'object' },
    _attributes:
     ObjectElement { _content: [Array], _storedElement: 'object' } },
  Annotation {
    _content: 'Missing required property: type',
    _storedElement: 'annotation',
    _meta:
     ObjectElement { _content: [Array], _storedElement: 'object' },
    _attributes:
     ObjectElement { _content: [Array], _storedElement: 'object' } } ]
```

From that list you can see that it complains about "Additional properties not allowed: $schema", but it also behaves differently as compared to when it's run without the added `$schema` property in `makeParameter`. 

I suppose adding that `$schema` property influences how the parsing is done? @kylef @pksunkara could you please advise me on this? Thank you.

kylef commented:
@artem-zakharchenko you will need to add the entry into the schema generated at https://github.com/apiaryio/api-elements.js/blob/8cd312061148abad63610a3b75afbe9944df32a6/packages/fury-adapter-swagger/test/inherit-parameters-test.js#L231. Note that "parameter.schema" is NOT JSON Schema, it's Swagger Schema. It just happens to be that they were the same in this case; but that no longer applies.

```js
const parameterSchema = source.paths['/'].parameters[0].schema; // but actually make copy of it so we don't modify it
parameterSchema.$schema = '...';
const schema = JSON.stringify(parameterSchema);
```

artem-zakharchenko commented:
@kylef any objection to isolate that logic into a helper utility?

```js
const schema = getSchemaFromSource(source)

function getSchemaFromSource(source) {
  const parameterSchema = _.clone(source)
  parameterSchema.$schema = '...'
  return JSON.stringify(oarameterSchema)
}
```

We're going to use this quite often in this test.

kylef commented:
The code in particular is deriving a JSON Schema from the Parameter's Swagger Schema and in doing so it will add the `$schema` now.

---

Could you add a part to `packages/fury-adapter-swagger/CHANGELOG.md` to specify the changes too, we have changelogs in this repo for each package:

```
## TBD

### Bug Fixes

- <some description of the problem we are solving>
```

artem-zakharchenko commented:
Should I squash updated fixtures commit into the `fix` one? Not sure how you approach this in the repo.

kylef commented:
> Should I squash updated fixtures commit into the fix one? Not sure how you approach this in the repo.

Yeah; although we can do that during merge (squash and merge in GH). I'd prefer to keep every commit passing tests so that tools like [`git bisect`](https://git-scm.com/docs/git-bisect) can be used. It makes it simpler if we need to do things like revert changes etc too.

kylef commented:
Changes look good :+1:, perhaps just adding changelog entry and we can merge this.

artem-zakharchenko commented:
> I'd prefer to keep every commit passing tests so that tools like git bisect can be used.

A valid point. I've squashed the commits and updated the relevant CHANGELOG. Please take a look. Thanks.
